File: android/app/src/main/java/com/ayuvo/health/models/ServingAmountExpression.kt
package com.ayuvo.health.models

import java.text.DecimalFormatSymbols
import java.util.Locale

object ServingAmountExpression {
    fun evaluate(expression: String, locale: Locale = Locale.getDefault()): Double? {
        val value = Parser(expression, locale).parse() ?: return null
        return value.takeIf { it.isFinite() }
    }

    fun appendOperator(expression: String, operator: Char): String {
        if (!isOperator(operator)) return expression
        val trimmed = expression.trim()
        if (trimmed.isEmpty()) return expression
        return if (isOperator(trimmed.last())) trimmed.dropLast(1) + operator else trimmed + operator
    }

    fun containsOperation(expression: String): Boolean =
        expression.withIndex().any { (index, character) -> index > 0 && isOperator(character) }

    private fun isOperator(character: Char): Boolean = character in "+-−*×/÷"

    private class Parser(expression: String, locale: Locale) {
        private val input = expression
            .replace('−', '-')
            .replace('×', '*')
            .replace('÷', '/')
        private val decimalSeparators = buildSet {
            add('.')
            add(DecimalFormatSymbols.getInstance(locale).decimalSeparator)
        }
        private var index = 0

        fun parse(): Double? {
            skipWhitespace()
            if (index >= input.length) return null
            val value = parseExpression() ?: return null
            skipWhitespace()
            return value.takeIf { index == input.length }
        }

        private fun parseExpression(): Double? {
            var value = parseTerm() ?: return null
            while (true) {
                skipWhitespace()
                val operation = peek()
                if (operation != '+' && operation != '-') return value
                index += 1
                val rhs = parseTerm() ?: return null
                value = if (operation == '+') value + rhs else value - rhs
            }
        }

        private fun parseTerm(): Double? {
            var value = parseFactor() ?: return null
            while (true) {
                skipWhitespace()
                val operation = peek()
                if (operation != '*' && operation != '/') return value
                index += 1
                val rhs = parseFactor() ?: return null
                if (operation == '/') {
                    if (kotlin.math.abs(rhs) <= Math.ulp(1.0)) return null
                    value /= rhs
                } else {
                    value *= rhs
                }
            }
        }

        private fun parseFactor(): Double? {
            skipWhitespace()
            if (peek() == '+') {
                index += 1
                return parseFactor()
            }
            if (peek() == '-') {
                index += 1
                return parseFactor()?.let { -it }
            }
            return parseNumber()
        }

        private fun parseNumber(): Double? {
            skipWhitespace()
            val start = index
            var foundDigit = false
            var foundSeparator = false
            val normalized = StringBuilder()

            while (true) {
                val character = peek() ?: break
                when {
                    character.isDigit() -> {
                        foundDigit = true
                        normalized.append(character)
                        index += 1
                    }
                    character in decimalSeparators && !foundSeparator -> {
                        foundSeparator = true
                        normalized.append('.')
                        index += 1
                    }
                    else -> break
                }
            }

            if (!foundDigit || index == start) return null
            return normalized.toString().toDoubleOrNull()
        }

        private fun skipWhitespace() {
            while (peek()?.isWhitespace() == true) index += 1
        }

        private fun peek(): Char? = input.getOrNull(index)
    }
}

File: ios/calorietracker/Models/ServingAmountExpression.swift
import Foundation

enum ServingAmountExpression {
    static func evaluate(_ expression: String, locale: Locale = .current) -> Double? {
        var parser = Parser(expression: expression, locale: locale)
        guard let value = parser.parse(), value.isFinite else { return nil }
        return value
    }

    static func appending(_ operation: Character, to expression: String) -> String {
        guard isOperator(operation) else { return expression }
        var trimmed = expression.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return expression }

        if let last = trimmed.last, isOperator(last) {
            trimmed.removeLast()
        }
        return trimmed + String(operation)
    }

    static func containsOperation(_ expression: String) -> Bool {
        expression.enumerated().contains { index, character in
            isOperator(character) && index > 0
        }
    }

    private static func isOperator(_ character: Character) -> Bool {
        "+-−*×/÷".contains(character)
    }

    private struct Parser {
        private let characters: [Character]
        private let decimalSeparators: Set<Character>
        private var index = 0

        init(expression: String, locale: Locale) {
            self.characters = Array(
                expression
                    .replacingOccurrences(of: "−", with: "-")
                    .replacingOccurrences(of: "×", with: "*")
                    .replacingOccurrences(of: "÷", with: "/")
            )
            var separators: Set<Character> = ["."]
            if let separator = locale.decimalSeparator?.first {
                separators.insert(separator)
            }
            self.decimalSeparators = separators
        }

        mutating func parse() -> Double? {
            skipWhitespace()
            guard index < characters.count, let value = parseExpression() else { return nil }
            skipWhitespace()
            return index == characters.count ? value : nil
        }

        private mutating func parseExpression() -> Double? {
            guard var value = parseTerm() else { return nil }
            while true {
                skipWhitespace()
                guard let operation = peek(), operation == "+" || operation == "-" else { return value }
                index += 1
                guard let rhs = parseTerm() else { return nil }
                value = operation == "+" ? value + rhs : value - rhs
            }
        }

        private mutating func parseTerm() -> Double? {
            guard var value = parseFactor() else { return nil }
            while true {
                skipWhitespace()
                guard let operation = peek(), operation == "*" || operation == "/" else { return value }
                index += 1
                guard let rhs = parseFactor() else { return nil }
                if operation == "/" {
                    guard abs(rhs) > .ulpOfOne else { return nil }
                    value /= rhs
                } else {
                    value *= rhs
                }
            }
        }

        private mutating func parseFactor() -> Double? {
            skipWhitespace()
            if peek() == "+" {
                index += 1
                return parseFactor()
            }
            if peek() == "-" {
                index += 1
                return parseFactor().map { -$0 }
            }
            return parseNumber()
        }

        private mutating func parseNumber() -> Double? {
            skipWhitespace()
            let start = index
            var foundDigit = false
            var foundSeparator = false
            var normalized = ""

            while let character = peek() {
                if character.isNumber {
                    foundDigit = true
                    normalized.append(character)
                    index += 1
                } else if decimalSeparators.contains(character), !foundSeparator {
                    foundSeparator = true
                    normalized.append(".")
                    index += 1
                } else {
                    break
                }
            }

            guard foundDigit, index > start else { return nil }
            return Double(normalized)
        }

        private mutating func skipWhitespace() {
            while let character = peek(), character.isWhitespace {
                index += 1
            }
        }

        private func peek() -> Character? {
            characters.indices.contains(index) ? characters[index] : nil
        }
    }
}
